### PR TITLE
feat(storybook): add Webpack 5 support for React Storybook setups

### LINKED ISSUE #6703: React "build-storybook" fails in v12.7.0 and Webpack5
## Current Behavior
React `build-storybook` fails in v12.7.0 & Webpack 5.

## Expected Behavior
Should build without errors

## Steps to Reproduce
* Create a repo with v12.7.0 and React app
* Generate storybook configuration `nx g @nrwl/react:storybook-configuration`
* run `nx run project:build-storybook`

### Failure Logs
```sh
>  NX  Storybook builder starting ...

info => Cleaning outputDir: /Users/ddd/frontend/dist/storybook/foo
info => Loading presets
info => Compiling manager..
info => Compiling preview..
info => Loading 1 config file in "/Users/ddd/frontend/apps/foo/.storybook"
info => Loading 9 other files in "/Users/ddd/frontend/apps/foo/.storybook"
info => Adding stories defined in "/Users/ddd/frontend/apps/foo/.storybook/main.js"
info => Loading Nrwl React Storybook preset from "@nrwl/react/plugins/storybook"

>  NX  Using webpack 5. Reason: detected version 5 in node_modules/webpack/package.json

info => Using default Webpack4 setup
Invalid value used in weak set
TypeError: Invalid value used in weak set
    at WeakSet.add (<anonymous>)
    at MiniCssExtractPlugin.apply (/Users/dan/work/frontend/node_modules/mini-css-extract-plugin/dist/index.js:362:18)
    at webpack (/Users/dan/work/frontend/node_modules/@storybook/builder-webpack4/node_modules/webpack/lib/webpack.js:51:13)
    at Object.build (/Users/dan/work/frontend/node_modules/@storybook/builder-webpack4/dist/cjs/index.js:179:18)
    at async Promise.all (index 1)
    at async buildStaticStandalone (/Users/dan/work/frontend/node_modules/@storybook/core-server/dist/cjs/build-static.js:141:28)

```

### Environment
```sh
  Node : 14.16.1
  OS   : darwin x64
  yarn : 2.4.2
  
  nx : Not Found
  @nrwl/angular : Not Found
  @nrwl/cli : 12.7.0
  @nrwl/cypress : 12.7.0
  @nrwl/devkit : 12.7.0
  @nrwl/eslint-plugin-nx : 12.7.0
  @nrwl/express : Not Found
  @nrwl/jest : 12.7.0
  @nrwl/linter : 12.7.0
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/nx-cloud : Not Found
  @nrwl/react : 12.7.0
  @nrwl/schematics : Not Found
  @nrwl/tao : 12.7.0
  @nrwl/web : 12.7.0
  @nrwl/workspace : 12.7.0
  @nrwl/storybook : 12.7.0
  @nrwl/gatsby : Not Found
  typescript : 4.3.5

```

### PR DESCRIPTION
**Pull request by juristr**
## Current Behavior

Right now, [React with Webpack 5 is just an opt-in](https://nx.dev/latest/react/guides/webpack-5), which is also why the current Storybook generator doesn't use the Webpack 5 setup by default.

## Expected Behavior

This PR adds two things:

a) a warning if an existing Storybook configuration doesn't use the Webpack 5 configuration

image

b) Detects whether the React setup uses Webpack 5 and makes sure to generate proper Webpack 5 setup for Storybook react stories

#### User migrates to Webpack 5 

User uses [this guide](https://nx.dev/l/a/guides/webpack-5). 
When they run
```
npx nx g @nrwl/web:webpack5
```
to migrate their React projects, this generator will also do the following:
1. Check if they have the `@storybook/react` package installed
2. It will install `@storybook/builder-webpack5` and `@storybook/manager-webpack5` if they are not already installed
3. It will add the `builder: 'webpack5'` option in the Storybook configurations (`.storybook/main.js`) of their React projects

## Related Issue(s)

Fixes #6703

### CODE REVIEW

File: packages/storybook/src/generators/init/init.ts [hunk 1]
```diff
@@ -85,8 +85,23 @@ function checkDependenciesInstalled(host: Tree, schema: Schema) {
...
       devDependencies['@storybook/react'] = storybookVersion;
     }
-  }
 
+    if (isReactUsingWebpack5()) {
+      if (
+        !packageJson.dependencies['@storybook/builder-webpack5'] &&
+        !packageJson.devDependencies['@storybook/builder-webpack5']
+      ) {
+        devDependencies['@storybook/builder-webpack5'] = storybookVersion;
```
Comment (FrozenPandaz) [inline packages/storybook/src/generators/init/init.ts:94]: So when someone runs `nx g migrate-to-webpack5`, do they also get it setup for storybook? Or do they have to run `nx g storybook-configuration`/ `nx g @nrwl/storybook:init` after running `migrate-to-webpack5`?

Comment (juristr) [reply packages/storybook/src/generators/init/init.ts:94]: there's a `migrate-to-webpack5` generator? 

But no :). this PR basically adds the fact that you can generate new Storybook setups that use Webpack 5 if the workspace uses webpack 5 for the React apps. + older storybook setups will show you a warning how to upgrade.

The problem with the storybook migrations is always the messy potentially highly customized Storybook JS config files 🤷 .

But my idea is to maybe have an opt-in generator in an upcoming PR, that can be launched manually and which tries to migrate you for the setups "we understand" if that makes sense

File: packages/storybook/package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
   },
   "dependencies": {
     "@nrwl/cypress": "*",
+    "@nrwl/web": "*",
```
Comment (jaysoo) [inline packages/storybook/package.json:32]: Do we want to add dependency on `@nrwl/web` or just copy the parts needed from `webpack/entry`? I think angular workspaces probably won't need to have this package installed?

**Review by jaysoo — APPROVED** [review]

**Comment by jaysoo** [comment]
Need this for the webpack upgrade. Merging it and removing the `isWebpack5` checks :)

**Pull request merged**